### ISSUE TITLE
Protected non-virtual instance members

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#3825]
This is an attempt at a full feature specification for protected instance members, as another solution to #835.
# Dart protected instance members

Dart has only one notion of declared access restriction: Library privacy.

Having private names solves two problems: Avoiding internal implementations having name clashes with third-party code, and ensuring what third-party code is able, and not able, to access, which makes it easier to ensure invariants. 

Since the author controls the size and scope of a library, and is expected to be fully cooperative with all code they have the power to modify anyway, that’s a rather sweet spot of privacy. By scoping a library to a single class, or to what would be an entire directory of files in another language, the author can choose anything between class privacy and “package privacy”. It’s an accessibility where the author can control the scope.

One accessibility feature that other object oriented languages have, and that Dart cannot emulate with library privacy, is *protected access*: Instance members of a class which can only be accessed from subclasses which extend that class.

Protected members are useful for reusable libraries. A library can expose a base or skeleton class that third-party subclasses can extend to provide customized classes. However, if the subclass needs to interact with the base class, it can currently only do so if the base class exposes *public* API, which then contaminates the public API of the subclass with operations that end-users shouldn’t be using.

Dart provides the `@protected` annotation which makes the analyzer warn if a non-subclass invokes the annotated member. That’s not a safe solution, the API is still public and accessible, anyone can choose to ignore the warning.

## Instance protected members

We add instance members that can only be invoked by subclasses. Such a member can be declared in a `class`, `enum` or `mixin` declaration. It’s available to any class which has the `class` or `enum`, or an application of the `mixin`, in its superclass chain. This includes the class itself &mdash; and for enums, nothing else.

### Syntax

The immediate choice is to make `protected` a built-in identifier and use it as a modifier:

```dart
protected int foo = 42;
protected int bar(int x) => x;
```

Since protected members cannot be static or abstract, we just have to decide whether it goes before or after `external`. Probably after, mainly because `external` is currently always first. So, same place as `static`, which makes sense.

Let’s go with this, it can be changed if we have a better idea.

### Semantics

We ensure that the modifier is only allowed where it makes sense.

* It’s a compile-time error if any non-instance member is declared protected.
* It’s a compile-time error if a member declaration is both protected and abstract.

* A mixin application class of a mixin with a protected member counts as a class with a protected member declaration.

* It’s a compile-time error if an instance member is declared as protected, and the containing declaration is not a `class`, `mixin` or `enum` declaration. _(Not allowed inside `extension` or `extension type` declarations.)_
  * Protected members count the same as non-protected members when it comes to name clashes inside a declaration.
  * It’s a compiler-time error if a class declares a static member with the same name as a protected member of its superclass. _For now. It’s not a actually a problem, but it isn’t a problem for normal instance members either, so we stay consistent._

* It’s a compile-time error if an instance member declaration is marked as protected, and any superinterface of the containing declaration has a member with the same name. _If a member is already public, so it exists in the interface, it cannot be made protected._
  * If a subclass introduces a public signature for an inherited protected member implementation, the member becomes public for that class, can be invoked through the interface or in dynamic calls on instances that implement that subclass.

The behavior of a correctly declared protected member is:

* A protected member introduces an *implementation* to the containing class, but no interface member signature in the declaration’s interface.

  * _Normal declarations introduce both, and abstract declarations introduce only an interface member signature._

* A protected member cannot be invoked through dynamic invocation. *In a dynamic invocation, if an instance member is found which is not in the interface of the receiver’s runtime type, dynamic invocation treats it as an inaccessible member.*

* Protected members can only be invoked through `super` or `this` invocations (including implicit `this` invocations for plain identifiers that are not in the lexical scope).

  * If invoked through `super`, it works just like any other `super` member invocation, directly invoking the implementation of the member. _Super member invocations do not use the interface of the superclass, but directly invokes the implementation._

  * A `this`-based member access is changed as follows:

    * If the interface of the surrounding class has a member with the corresponding name, the invocation is a virtual member invocation as normal.
    * Otherwise if the class declares or inherits a corresponding protected member declaration, that declaration is directly invoked, similar to a `super`-invocation.
    * Otherwise check for extension members as normal.

    *That is, one extra step is inserted between checking the interface and falling back on extensions.*

* Mixins declaring protected members can invoke those members using `this` invocations, but not `super` invocations. _(Until Dart gets composite mixins, a super-invocation of a mixin is effectively an interface member invocation)_.

## Summary

Protected members are implementation that is not exposed in the implicit interface of a class. They cannot be abstract, which means being interface without implementation. Like other implementation, a protected member is only inherited through the *superclass* relation, and can be invoked normally using `super` member invocations.

Subclasses can introduce the an interface signature for the protected member, making it public API (or library private, if named with a private name).

Typed invocations cannot see protected members, since they are not in the type’s interface. Dynamic invocations treat protected members that haven’t been made public as inaccessible.

Protected members can be invoked using `this.name`. If the name is not in the surrounding class’s interface, but an implementation is inherited, that implementation (statically resolved) is invoked. 

Protected members are non-virtual, non-interface methods. They can be shadowed by other protected or non-protected members. They become virtual only when made public by being given a interface signature, without introducing new implementation.

**Comment by escamoteur** [comment]
>Protected members are non-virtual, non-interface methods. They can be shadowed by other protected or non-protected members. They become virtual only when made public by being given a interface signature, without introducing new implementation.

I don't really understand this paragraph. why can they be shadowed? and why aren't they virtual? 

```dart
class A {
  protected void protectedFunc(){
     print('in A');
  };
  void init(){
    protectedFunc();
  }
}

class B extends A{
  @override
  protected void protectedFunc(){
    print('in B');
  }
}

B b;
b.init(); // -> should call B.protectedFunc() and not the one of A
```

**Comment by lrhn** [comment]
A virtual method is late-bound, the actual implementation chosen based on the runtime type of the receiver it's called on. An interface method is basically the same in Dart. (Java has separate JVM instructions for interface invocations and virtual instance member invocations because they have separate `interface` and `class` declarations. It's all interfaces in Dart.)

Now, the question is whether this is the *correct* design. Your example would print `in A` the way this is specified, because `A` doesn't know that `protectedFunc` is overridden in subclasses. But that doesn't seem *necessary*, or particularly useful.
If a call could be virtual, that is: a `this.name` invocation would be virtual, and a `super.name` is still not, then subclasses *can* override protected members (and must if the protected members is abstract, which now also makes sense).

That's actually much more useful. (Aka: Doh! Why didn't I see that.)

I'll see when I can find time do an update :)

**Comment by escamoteur** [comment]
No, it should print `in B` because the runtime type of `b` is B

**Comment by lrhn** [comment]
I agree that it's a good idea if it prints `B`, but that was not what I had specified. As specified, it would print `A`.
So I fixed it. :wink:  (And now nobody can see the first version any more, and laugh at its naivity. Hah!)

**Comment by mateusfccp** [comment]
I know this proposal is not directly related to this, but if we are ever going to have `protected` members, we should probably want to deal with `private`.

As you yourself say, similar things should look similar. Having the `_` prefix meaning `private` and an actual `protected` prefix will probably look too asymmetrical, and also be confusing to people non-familiar with the language (IMO, the `_` prefix is already confusing).

**Comment by escamoteur** [comment]
I would love to see that. Especially as it looks totally ugly to use _properties as constructor and there is even a lint discouraging it so you need initilizers.
Am 22. Mai 2024, 20:30 +0200 schrieb Mateus Felipe C. C. Pinto ***@***.***>:

**Comment by tatumizer** [comment]
There are different scopes of privacy. "protected" is also kind-of-private, but it's private for the class and all the descendants.
The declaration `int _foo` in a class is private to the library, not to the class. If `private x` is introduced, then it must be private to the class (not visible even to descendants), so it doesn't have the same visibility as `_x`. Both kinds have their uses.

Does dart have a concept of package-private sub-packages (libraries)? The fact that the libraries in `/src` directory can be accessed from the outside via `import 'package:name/src/foo.dart'` breaks their assumed "privacy". We can also envision package-private variables, functions, methods, classes etc. Do you want to support all of those? If not, why not?

(is `/src` directory treated as a sub-package? Its ontological status is unclear.)

Regardless, I find `_foo` syntax for library-private variables very convenient. These variables are so common that special-casing their syntax is well-justified IMO. Dart won't be the same language without this convention.

**Comment by escamoteur** [comment]
@tatumizer I agree, that it would make sense to have `int _foo` as well as `private int foo` where the later one would really mean class private.  
For my understanding library private means accessible to everything inside one file/part-of files, but not from out side of that. There is no concept of a package as a library as such.

**Comment by lrhn** [comment]
A prefix `private` to declare a library private declaration could work for any statically resolved declaration. It would simply not be seen from any other library. It might cause more name collisions to use non-`_` names, but you could still use those of you wanted to. 

The reason to use a name prefix for library privacy is instance members, and dynamic invocation in particular. 

If two classes in a superclass chain both introduce a private name, then the two are considered different names. 
If we used `private` on the declarations instead, then they should still be considered different names, but I've cannot see that at the invocation.
If a superclass in the same library has a `private int foo;` field, and another superclass from another library adds a public `void foo()` method, then either
- there is a conflict, which there wouldn't be today, making you want to prefix your private instance variables with your library name, or something similarly cumbersome, or
- the subclass now cannot reference both of these, because to it, they have the same name, `this.foo` can only regret to one. (That's obviously fixed by renaming the library private variable, but again, you start wanting to prefix your library private variables, to avoid such churn.)

For `dynamic` invocation, it would need to know the library it's called from, to know which `foo` it should find out skip. That's doable, but also error prone.

With the leading `_`, one can see at the user-site that the library's private is wanted.

For `protected`, which is cross library, using a modifier is easier, because it prohibits most of the problem cases.
It could also use `%id` as name, and I'm sure it would work. There'd be no way to make a protected member public, would just have to forward to it instead.

But all the good name prefixes are taken.

**Comment by munificent** [comment]
> * A protected member is considered _inaccessible_ when accessed through a typed member invocation for any receiver other than `this`. _Doing `Foo o = ...; o.id();` where `Foo`'s interface has a protected member named `id`, is a compile-time error, because `Foo` does not have an _accessible_ member named `id`._

This means that you're proposing doing "instance protected" versus "class protected" like other OOP languages I'm familiar with do (C++, C#, Java). Is that a deliberate choice? For what it's worth, I have often found it useful to be able to call protected members on other instances of the surrounding class. We essentially have that capability in Dart with *private* members because of library privacy (although not up the superclass chain of course). So it feels a little strange that you could call a *private* method on some other instance of your own class, but not a *protected* method on that same instance.

(This also raises the question of whether you can have private + protected methods and whether that is a useful combination.)

**Comment by lrhn** [comment]
> "instance protected" ... Is that a deliberate choice?

Yes. I have always believed that to be a better match for Dart because of two things: dynamic invocations and not having interface decelerations.

Class based `protected` makes more sense when you can see at the call point whether you are calling the protected member or not. You cannot see that for a dynamic invocation. 

The alternative would be to just make protected members inaccessible in dynamic invocations, which they are in this design anyway. It just feels like a missing feature then, not a logical design consequence. 

Being "inside a subclass" is not the same protection if one can simply `implements Superclass` or an `abstract ... extends Superclass` and add a static member to it that accesses the protected members of `Superclass` that your class is extending.
If you extend a superclass with protected members, someone should need to extend your class to be able to access them on your objects. 

This way they are more like implantation members that are not part of the interface, which can still be accessed on `this` and `super`, but not through an interface member invocation. Like the opposite of an abstract member, which adds to the interface, but has no implementation.

It does mean no protected statics. Again, when you can implement any class, class-based protected statics it's not much of a protection.
